tests/test_algorithm_learning.py

```diff
@@ -0,0 +1,355 @@
+from reinforcement.algorithm.reinforce import Reinforce
+
+
+class Space:
+    pass
+
+
+class Discrete(Space):
+    def __init__(self, n):
+        self.n = n
+
+    def sample(self):
+        return np.random.randint(self.n)
+
+    def __repr__(self):
+        return f"Discrete(n={self.n})"
+
+
+class Box(Space):
+    def __init__(self, shape):
+        self.shape = shape
+
+    def __repr__(self):
+        return f"Box(shape={self.shape})"
+
+
+class BinaryBox(Box):
+    def __init__(self):
+        super().__init__(shape=(2,))
+
+    @staticmethod
+    def sample():
+        r = np.random.randint(1)
+        return np.array([r, 1 - r])
+
+
+class NormalDistribution:
+    def __init__(self, mean, std):
+        self.mean = mean
+        self.std = std
+
+    def sample(self):
+        return np.random.normal(self.mean, self.std)
+
+
+class SwitchingMDP:
+    def __init__(self):
+        self.action_space = Discrete(2)
+        self.observation_space = BinaryBox()
+        self.len_episode = 100
+        self._reward = NormalDistribution(1, 0.1)
+        self._penalty = NormalDistribution(-1, 0.1)
+        self.avg_max_reward = self.len_episode * self._reward.mean
+        self._current_state = None
+        self._episode = None
+
+    def reset(self):
+        self._episode = 0
+        self._current_state = self.observation_space.sample()
+        return self._current_state
+
+    def step(self, action):
+        self._episode += 1
+        if self._is_valid(action):
+            self._flip_state()
+            return self._current_state, self._reward.sample(), self._is_done(), None
+        else:
+            return self._current_state, self._penalty.sample(), self._is_done(), None
+
+    def _is_valid(self, action):
+        return action is not None and self._current_state[0] != action
+
+    def _flip_state(self):
+        self._current_state = np.array([self._current_state[1], self._current_state[0]])
+
+    def _is_done(self):
+        return self._episode == self.len_episode
+
+    def __repr__(self):
+        return f"SimpleDeterministic(action_space={repr(self.action_space)}, " \
+               f"observation_space={repr(self.observation_space)}, len_episode={self.len_episode}," \
+               f"avg_total_return={repr(self.avg_max_reward)})"
+
+
+class RandomAgent(AgentInterface):
+    def __init__(self, action_space):
+        self._action_space = action_space
+
+    def next_action(self, _):
+        return self._action_space.sample()
+
+    def signal(self, _):
+        pass
+
+    def train(self):
+        pass
+
+
+class OptimalAgent(AgentInterface):
+    def next_action(self, observation):
+        return 1 - observation[0]
+
+    def signal(self, reward):
+        pass
+
+    def train(self):
+        pass
+
+
+class LinearPolicy:
+    def __init__(self, plotter, lr=1):
+        self._lr = lr
+        self._m = np.random.normal(1, 0.001)  # high m avoids local maxima - use negative m for demonstration
```

Comment:
```suggestion
        self._m = np.random.normal(1, 0.001)  # high slope avoids local maxima - use negative m for demonstration
```
maybe refer to m as slope in the comment to make it more clear